### PR TITLE
Calculate amounts EPP

### PR DESCRIPTION
**Pull request by Mropat**
### This PR adds | fixes:
- New calculate amount ng EPP

### Review:
- [ ] Code approved by
- [ ] Tests executed by
- [ ] "Merge and deploy" approved by

This [version](https://semver.org/) is a:
- [ ] **MINOR** - when you add functionality in a backwards compatible manner

### CODE REVIEW

File: cg_lims/EPPs/udf/calculate/calculate_amount_ng.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+from cg_lims.get.artifacts import get_qc_output_artifacts
+from cg_lims.get.samples import get_one_sample_from_artifact
+
+LOG = logging.getLogger(__name__)
+
+
+@click.command()
+@click.pass_context
+def calculate_amount_ng(ctx):
+    """Calculates amount of samples for qc validation."""
```
Comment (mayabrandi) [inline cg_lims/EPPs/udf/calculate/calculate_amount_ng.py:16]:
```suggestion
    """Calculates amount of samples for qc validation.
    
    Reading and writing qc measurement udfs only."""
```

Comment (mayabrandi) [reply cg_lims/EPPs/udf/calculate/calculate_amount_ng.py:16]: If you can, folmulate this in a bettr way than i did. I just vant to inform that this EPP only works on messurements - not analytes

File: cg_lims/EPPs/udf/calculate/calculate_amount_ng.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        if missing_udfs_count:
+            raise MissingUDFsError(
+                f"Udf missing for {missing_udfs_count} sample(s): {','.join(samples_with_missing_udf)}."
+            )
+        message = "Amounts have been calculated for all samples."
+        LOG.info(message)
+        click.echo(message)
+    except LimsError as e:
+        LOG.error(e.message)
+        sys.exit(e.message)
```
Comment (mayabrandi) [inline cg_lims/EPPs/udf/calculate/calculate_amount_ng.py:47]:
```suggestion
        sys.exit(e.message)
        
```

**Review by mayabrandi — COMMENTED** [review]
Great Maria! 

You will have to add the the new EPP to the calcultate base cli
https://github.com/Clinical-Genomics/cg_lims/blob/2ad1d92d4e68c4f94d846cd4b3a4be382f44e175/cg_lims/EPPs/udf/calculate/base.py

It might be that we want this epp to be able to calculate amount on non-qc steps. Ie read and write to derived samples (not messuements). But I suggesty we do that change if/when it gets needed in future.

File: cg_lims/EPPs/udf/calculate/calculate_amount_ng.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    process = ctx.obj["process"]
+    lims = ctx.obj["lims"]
+
+    try:
+        artifacts = get_qc_output_artifacts(lims=lims, process=process)
+        missing_udfs_count = 0
+        samples_with_missing_udf = []
+        for artifact in artifacts:
+            sample = get_one_sample_from_artifact(artifact)
+            vol = artifact.udf.get("Volume (ul)")
```
[the remark this answers is not in the document]
Comment (Mropat) [reply cg_lims/EPPs/udf/calculate/calculate_amount_ng.py]: Done

File: cg_lims/options.py [hunk 4]
```diff
@@ -1,55 +1,111 @@
...
+        "-n",
+        "--process-type",
+        required=True,
+        multiple=True,
+        help=help,
     )
+
+
+def amount_keyword_option(help: str = "String of UDF used to get amount value") -> click.option:
+    return click.option("--amount-keyword", required=True, help=help)
```
Comment (mayabrandi) [inline cg_lims/options.py:101]:
```suggestion
    return click.option("--amount-udf", required=True, help=help)
```

File: cg_lims/options.py [hunk 5]
```diff
@@ -1,55 +1,111 @@
...
+
+
+def volume_keyword_option(help: str = "String of UDF used to get volume value") -> click.option:
+    return click.option("--volume-keyword", required=True, help=help)
```
Comment (mayabrandi) [inline cg_lims/options.py:105]:
```suggestion
    return click.option("--volume-udf", required=True, help=help)
```

File: cg_lims/options.py [hunk 6]
```diff
@@ -1,55 +1,111 @@
...
+
+
+def concentration_keyword_option(
+    help: str = "String of UDF used to get concentration value",
+) -> click.option:
+    return click.option("--concentration-keyword", required=True, help=help)
```
Comment (mayabrandi) [inline cg_lims/options.py:111]:
```suggestion
    return click.option("--concentration-udf", required=True, help=help)
```

**Review by mayabrandi — APPROVED** [review]
Suggest the option flaggs indicate they point to udfs

File: cg_lims/EPPs/udf/calculate/calculate_amount_ng.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mayabrandi) [reply cg_lims/EPPs/udf/calculate/calculate_amount_ng.py]: sweet

**Comment by mayabrandi** [comment]
I tested the script locally:

Skärmavbild 2021-01-29 kl  12 59 31

Skärmavbild 2021-01-29 kl  12 59 25

👍

**Comment by mayabrandi** [comment]
So nice to have you writing EPPs @Mropat !

File: cg_lims/EPPs/udf/calculate/calculate_amount_ng.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+LOG = logging.getLogger(__name__)
+
+
+@click.command()
+@options.concentration_keyword_option()
+@options.amount_keyword_option()
+@options.volume_keyword_option()
+@click.pass_context
+def calculate_amount_ng(
+    ctx: click.Context, amount_keyword: str, volume_keyword: str, concentration_keyword: str
```
Comment (Mropat) [inline cg_lims/EPPs/udf/calculate/calculate_amount_ng.py:20]:
```suggestion
    ctx: click.Context, amount_udf: str, volume_udf: str, concentration_udf: str
```

**Pull request merged**